Sources/SwiftSyntax/SyntaxData.swift

```diff
@@ -327,43 +331,52 @@ struct SyntaxData {
   ///   - index: The index pointing to where in the raw layout to place this
   ///            child.
   ///   - newChild: The raw syntax for the new child to replace.
+  ///   - newChildArena: The arena in which `newChild` resides.
   ///   - arena: The arena in which the new node will be allocated.
   /// - Returns: The new root node created by this operation, and the new child
   ///            syntax data.
   /// - SeeAlso: replacingSelf(_:)
-  func replacingChild(at index: Int, with newChild: RawSyntax?, arena: SyntaxArena) -> SyntaxData {
-    let newRaw = raw.layoutView!.replacingChild(at: index, with: newChild, arena: arena)
-    return replacingSelf(newRaw, arena: arena)
+  func replacingChild(at index: Int, with newChild: RawSyntax?, rawNodeArena: SyntaxArena?, allocationArena: SyntaxArena) -> SyntaxData {
+    if let newChild {
+      precondition(newChild.arena === rawNodeArena)
+    }
+    // After newRaw has been allocated in `allocationArena`, `rawNodeArena` will
+    // be a child arena of `allocationArena` and thus, `allocationArena` will
+    // keep `newChild` alive.
+    let newRaw = withExtendedLifetime(rawNodeArena) {
+      raw.layoutView!.replacingChild(at: index, with: newChild, arena: allocationArena)
+    }
+    return replacingSelf(newRaw, rawNodeArena: allocationArena, allocationArena: allocationArena)
   }
 
   /// Identical to `replacingChild(at: Int, with: RawSyntax?, arena: SyntaxArena)`
   /// that ensures that the arena of`newChild` doesn’t get de-allocated before
   /// `newChild` has been addded to the result.
   func replacingChild(at index: Int, with newChild: SyntaxData?, arena: SyntaxArena) -> SyntaxData {
```

Comment:
I still find this odd FWIW. I thought the idea behind `RawSyntax` was to avoid the retain, but using `SyntaxData` everywhere should remove that gain. Though I guess `replacingSelf` still just passes `RawSyntax`, so it'd only be a single retain vs one for each parent?